apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryDataReader.cs

```diff
@@ -14,197 +14,207 @@
 
         /// <inheritdoc />
-        public override object this[int ordinal]
-        {
-            get { throw new NotImplementedException(); }
-        }
+        public override short GetInt16(int i) => GetFieldValue<short>(i);
 
         /// <inheritdoc />
-        public override object this[string name]
-        {
-            get { throw new NotImplementedException(); }
-        }
+        public override int GetInt32(int i) => GetFieldValue<int>(i);
 
         /// <inheritdoc />
-        public override bool HasRows { get; }
+        public override long GetInt64(int i) => GetFieldValue<long>(i);
+
+        /// <inheritdoc />
+        public override string GetName(int i) => Current?.Schema.Fields[i].Name;
 
         /// <inheritdoc />
         public override int GetOrdinal(string name)
         {
-            throw new NotImplementedException();
+            var fields = Current?.Schema.Fields;
+            if (fields == null)
+                return -1;
+
+            var index = 0;
+            foreach (var field in fields)
+            {
+                if (field.Name == name)
+                    return index;
+                index++;
+            }
+            return -1;
         }
 
         /// <inheritdoc />
-        public override string GetDataTypeName(int ordinal)
-        {
-            throw new NotImplementedException();
-        }
+        public override string GetString(int i) => GetFieldValue<string>(i);
 
         /// <inheritdoc />
-        public override Type GetFieldType(int ordinal)
+        public override object GetValue(int i) => this[i];
+
+        /// <inheritdoc />
+        public override int GetValues(object[] values)
         {
-            throw new NotImplementedException();
+            var count = Math.Min(FieldCount, values.Length);
+            for (var i = 0; i < count; i++)
+                values[i] = this[i];
+
+            return count;
         }
 
+        /// <summary>
+        /// </summary>
+        public override bool IsDBNull(int ordinal) => GetValue(ordinal) == null;
+
+        /// <inheritdoc />
+        public override int FieldCount => Current?.Schema.Fields.Count ?? 0;
+
         /// <inheritdoc />
-        public override object GetValue(int ordinal)
+        protected override void Dispose(bool disposing)
         {
-            throw new NotImplementedException();
+            _results?.Dispose();
+            _connectionToClose?.Close();
+            base.Dispose(disposing);
         }
 
-        /// <inheritdoc />
-        public override IEnumerator GetEnumerator()
+        private BigQueryRow Current => _isEnumerating ? _results?.Current : null;
+
+        private BigQueryResults BigQueryResults { get; }
+
+        private Task<bool> MoveNextAsync() => GetResultsAsync().MoveNext();
+
+        private IAsyncEnumerator<BigQueryRow> GetResultsAsync()
         {
-            throw new NotImplementedException();
+            if (_results != null)
```

Comment:
As far as I remember, GetRowsAsync doesn't immediately make any calls - so perhaps it would be simpler to do this eagerly on construction? I believe we only start fetching data when MoveNext() is falled. Probably worth checking though.

Reply:
But HasRows uses the existence of Current for its value.  So it might make code a little cleaner, but wouldnt help with predetermining HasRows.

Reply:
I've fixed this up in the latest commit.